### ISSUE TITLE
Sync with NSFetchedResultsController

### ISSUE THREAD

**Issue by Gerungofulus** [hyperoslo/Sync#112]
Hi there! It's me again :) 
I definitely want to write your documentation, but right now my code starts falling apart. So here is the problem:

I have the following CoreData Model set up:
![bildschirmfoto 2015-07-17 um 19 40 29](https://cloud.githubusercontent.com/assets/9848525/8753394/c47a6c4e-2cbb-11e5-9c06-36e38a94582e.png)

Now I want to add a picture locally by doing the following (I am using old code that I wrote when I did not use Sync yet):

So I basically get an image, that I want to add to the Photo Items

```swift
func gotimage(image:UIImage){
  let col = GrocliCollectionManager.sharedInstance.getDefaultWriteCollection()
  
  let uid = createUniqueFileUIDForCollection(col)
  //create subdir if needed
  GrocliFileHandler.sharedInstance.createFolderForCollection(col)
  let filename = (col.remoteID + "/" + uid + ".jpg")
  let ok = self.writeImage(image, to: filename)
  
  
  if(ok){
      let photoDic = [
          "id" : uid,
          "collectionId" : col.remoteID,
          "fileurl"   : filename,
          "isBought"  : false
      ]
      var photo = GrocliCollectionManager.sharedInstance.createNonExistingPhoto(photoDic)
      
      
      if(photo != nil){
          photo?.isDownloaded = true
          photo?.isUploaded = false
          col.photos = col.photos.setByAddingObject(photo!)
          photo?.collectionId = col.remoteID
          //println(GrocliFileHandler.sharedInstance.createFileName(photo!))
          var error:NSError?
          //photo?.managedObjectContext?.save(&error)
          (UIApplication.sharedApplication().delegate as! AppDelegate).dataStack?.persistWithCompletion({ () -> Void in
              if(GrocliLoginHandler.sharedInstance.loggedIn as Bool){
                  GrocliAPIHandler.sharedInstance.uploadPicture(photo!)
              }
          })
          if(error != nil){
              GrocliErrorHandler.handleCoreDataSaveError(error!)
          }
          
      }
  }
}
```

The `createNonExistingPhoto(dict:NSdictionary)` function basically just take a JSON array and creates a CoreDAta PhotoItem on the `AppDelegate's Datastack's mainContext`.

You see, that I tried different version to save and tested those. They work. Now the uploading takes place in the `APIHandler's uploadPicture`. When the uploading was successful, I get a server response which contains the new ID for the photo which I need to set. I do this in the `completionHandler`

```swift
if let success: AnyObject = jsonDictionary!["success"]{
                      if(success as! Bool){
                          println(jsonDictionary)
                          if let payload = jsonDictionary!["payload"] as? NSDictionary{
                              println(payload)
                              let file = payload["file"] as! NSDictionary
                              photo.remoteID = file["id"] as! String
                              photo.isUploaded = true
                              
                          }
                          var err:NSError?
                          //photo.managedObjectContext?.save(&err)
                          
                          (UIApplication.sharedApplication().delegate as! AppDelegate).dataStack!.persistWithCompletion({ () -> Void in
                              if(err != nil){
                                  GrocliErrorHandler.handleCoreDataSaveError(err!)
                              }else{
                                  //self.getCollections()
                              }
                          })
                          
                          
                      }
                      
                  }
```

For example I got a file Dictionary that looked likes this
   ` file =     {
        fileurl = "dd432a4135acb80e/D4CE12234E9A4AFEB5FD.jpg";
        id = 1a309871d7618851;
        isBought = 0;
    };`
Again this does work. I can successfully save this photo. Now I use a NSFetchedResultsController on the Photos that I set up like this:

```swift
func photoItemFetchRequest()->NSFetchRequest{
    
    var request = NSFetchRequest(entityName: "GrocliPhotoItem")
    
    request.predicate = NSPredicate(format: "deletedOnServer == NO && collection != nil", argumentArray: ["isDownloaded","deletedOnServer", "collection"])
    // TODO Gekaufte nicht anzeigen
    var sortDesc = NSSortDescriptor(key: "collection.remoteID", ascending: true)
    var ownColDesc = NSSortDescriptor(key: "collection.ownCollection", ascending: false)
    //var boughtSortDes = NSSortDescriptor(key: "isBought", ascending: true)
    request.sortDescriptors = [ownColDesc,sortDesc]
    return request
}

func getFetchedResultsController() -> NSFetchedResultsController {
    //TODO cache löschen wegtun
    NSFetchedResultsController.deleteCacheWithName("collectionCache")
    var fetchedResultsController = NSFetchedResultsController(fetchRequest: photoItemFetchRequest(), managedObjectContext: self.appDelegate!.dataStack!.mainContext, sectionNameKeyPath: "collection.remoteID", cacheName: "collectionCache")
    return fetchedResultsController
}
```

Like I said: everything seems to be working fine. If I now pull all images from the server I get the following json Response
      `[{
    collectionId = dd432a4135acb80e;
    fileurl = "dd432a4135acb80e/D4CE12234E9A4AFEB5FD.jpg";
    id = 1a309871d7618851;
    isBought = 0;
}, {
    collectionId = dd432a4135acb80e;
    fileurl = "dd432a4135acb80e/E2A376033B294FC5A005.jpg";
    id = 7b298fbb38ba3a65;
    isBought = 0;
}, {
    collectionId = dd432a4135acb80e;
    fileurl = "dd432a4135acb80e/CB26B84FC46241CEA84C.jpg";
    id = ae6c424bd4665b9c;
    isBought = 0;
}, {
    collectionId = dd432a4135acb80e;
    fileurl = "dd432a4135acb80e/D3113522D516495A9F5F.jpg";
    id = 719b416f0cc3fa70;
    isBought = 0;
}]` 
The ID and the filePath of the first item actually exist in the database since I save the file Dictionary (see above). I use sync in the following way

```swift
let firstItem = photoArray.first
    if (firstItem != nil && firstItem!["id"] != nil) {
        let id = firstItem!["collectionId"] as? String // find out which collection to update
        if(id != nil){
            
            if let collection = self.getCollection(id: id!){
                println(photoArray)
                Sync.changes(photoArray, inEntityNamed: "GrocliPhotoItem", parent: collection, dataStack: self.appDelegate!.dataStack!, completion: { (error) -> Void in
                    if(error == nil){
                        GrocliFileDownloader.sharedInstance.downloadAllFilesForCollection(collection)
                    }else{
                        GrocliErrorHandler.handleFilesSyncError(error)
                    }
                })
            }
        }
    }
```

This seems to work, too. 
Nonetheless the fetchedResultsController fails instantly afterwards with 
    erminating app due to uncaught exception 'NSRangeException', reason: '*** -[_PFArray objectAtIndex:]: index (3) beyond bounds (3)'

Is it possible, that Sync first updates and adds items and deletes items in a second save afterwards? Then the fetchedResultsController would call the controllerDidChangeContent function right after the first save and it would report more items that it really has when it saves all the deletes right afterwards. When it is then asked for the items that don't exist anymore, it crashes. Does that make sense? The other question then is: Why is the picture not updated? I suggest, that it does not delete/update properly, because if it is the very first item, my fetchedResultsController contains the same item two times. Thanks in advance! I hope everything is going well. 

Best regards,

Pascal

**Comment by 3lvis** [comment]
Hi Pascal, sorry for not following up this before, I've been out in vacation-mode :sunglasses: 

Were you able to fix this?

**Comment by Gerungofulus** [comment]
Yeah man, vacations are important, I hope yours have been great! :)
I am really sorry to say, that I could not fix this problem. Do you have any idea what went wrong?

**Comment by 3lvis** [comment]
@Gerungofulus I haven't had anything like this happen to me before, sorry I can't be more helpful.

**Comment by Gerungofulus** [comment]
I won't have any time to spent on this particular issue within the next two weeks, but I'll investigate it later on and keep you updated

**Comment by 3lvis** [comment]
@Gerungofulus Don't worry, good luck :)

**Comment by ashkan18** [comment]
we had similar problem on our side, for us the issue was faulted objects which our fetched result controller couldn't access. We ended up following this which fixed our issue:

http://stackoverflow.com/questions/3923826/nsfetchedresultscontroller-with-predicate-ignores-changes-merged-from-different/3927811#3927811

I hope this helps.

**Comment by altyus** [comment]
what ^ said

**Comment by 3lvis** [comment]
@ashkan18 @altyus Thanks guys for pointing that you both had a similar problem.

I'm trying to figure out how to reproduce this bug.

Did you also had the problem in this method?

https://github.com/hyperoslo/Sync/blob/master/Source/Sync.h#L18-L22

**Comment by 3lvis** [comment]
@Gerungofulus I'm trying to reproduce this issue. Just to clear things up.

1.- You create a photo item with general information in the main context
2.- You upload this photo to the backend (in a background thread?) and update the photo created in step 1 with the new information that you get from the backend.
3.- You retrieve all the photo items for the current collection.
4.- (Expected) All the photo items should be inserted/removed/updated including the one that you added on step 1.

Is this correct?

**Comment by Gerungofulus** [comment]
I'm going to test the solution given by @ashkan18 Thanks for your support! @3lvis Yes, this is correct. Thank to all of you for trying to solve the problems I was supposed to solve. Sadly I am not skilled enough yet to come up with a solution of my own. I should have some free time to spent on this issue next week, so I'll let you guys know whether it worked out!

**Comment by cyupa** [comment]
Me again :open_mouth: It seems to me that Sync has some issues with synchronising contexts. I'm also using it with a NSFetchedResultsController... and Storyboards.

1. I load some data
2. Setup a NSFetchedResultsController on the main thread
3. Fetch (even with huge delays) -> NO DATA

The funny thing is that it sometime works, if I loose myself through breakpoints. And data is loaded in the UITableViewController.

**Comment by 3lvis** [comment]
@cyupa this might be related to something different thought. https://github.com/3lvis/DATASource/issues/32#issuecomment-150020808 ?

**Comment by cyupa** [comment]
I did some rework on DATAStack (still need to learn Swift - https://github.com/cyupa/DATAStack/blob/master/Source/DATAStack.swift) and I was able to get the data into the UI.
I changed the context mapping to:
writerContext - mainContext - backgroundContext
Where writerContext is parent for main and main is parent for backgroundContext.
They all share the same persistent store coordinator via writerContext.

The problem now is that after I load the data on the first launch and everything is ok, if I kill the app and relaunch, it does not load the SQLite file. I need to check how the persistentStoreCoordinator is set because it has a really complicated getter.
